src/NServiceBus.Extensions.Hosting/NServiceBusHostedService.cs

```diff
@@ -16,31 +16,45 @@ public NServiceBusHostedService(IStartableEndpointWithExternallyManagedContainer
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
-            try
+            // closure allocation is ok here
+            using (cancellationToken.Register(() => { endpointTcs.TrySetCanceled(); }, useSynchronizationContext: false))
             {
-                var endpoint = await startableEndpoint.Start(new ServiceProviderAdapter(serviceProvider))
-                    .ConfigureAwait(false);
-                endpointTcs.SetResult(endpoint);
-            }
-            catch (Exception e)
-            {
-                endpointTcs.SetException(e);
-                throw;
+                try
+                {
+                    var endpoint = await startableEndpoint.Start(new ServiceProviderAdapter(serviceProvider))
+                        .ConfigureAwait(false);
+                    endpointTcs.TrySetResult(endpoint);
+                }
+                catch (Exception e)
+                {
+                    endpointTcs.TrySetException(e);
+                    throw;
+                }
             }
         }
 
         public async Task StopAsync(CancellationToken cancellationToken)
         {
-            var endpoint = await endpointTcs.Task.ConfigureAwait(false);
-            await endpoint.Stop().ConfigureAwait(false);
+            // closure allocation is ok here
+            using (cancellationToken.Register(() => { endpointTcs.TrySetCanceled(); }, useSynchronizationContext: false))
+            {
+                var endpoint = await endpointTcs.Task.ConfigureAwait(false);
+                
+                var stopCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                using (cancellationToken.Register(() => { stopCompletionSource.TrySetCanceled(); }, useSynchronizationContext: false))
```

Comment:
I'm having a hard time understanding why we need a second TCS here, can we just use the token passed to Stop in WhenAny?

Reply:
Well if you find a `WhenAny` overload that accepts a token ;)

This commit is really just there if you want to also honor the token for the stop operation SLA.

Reply:
Given that our stop operation doesn't support cancellation, I'd vote to not use this approach.

Reply:
I'm not sure how the implementation detail of us not supporting cancellation should drive us to the conclusion of not honoring the SLAs of the host but I'll leave that up to you to decide. That's why I made this a separate commit that can be reverted if needed

Reply:
> I'm not sure how the implementation detail of us not supporting cancellation should drive us to the conclusion of not honoring the SLAs

imo it means that we should not just terminate the process while we haven't shut down properly because that's not something our shutdown process takes into account properly.

Reply:
Again take what you want from this PR by cherry picking, merging or even closing ;)

Reply:
will do, we're just figuring out the details